### PR TITLE
Implement new `exec` write-hook runner that will execute arbitrary binaries

### LINKED ISSUE #1275: Support for non-Python post-write hooks
**Describe the bug**
I'm using [Ruff](https://beta.ruff.rs/docs/) to lint and reformat my code, and I'm trying to get it working as a post-write hook in Alembic, but Ruff is actually written in Rust and is installed as a compiled binary, so it isn't included in `importlib` metadata as a console-scripts entrypoint, and Alembic throws an "Could not find entrypoint console_scripts.ruff" error.

**Expected behavior**
There should be an out of the box write hook runner that can execute arbitrary binaries, and not just the `console-scripts` runner that only supports Python modules' entrypoints.

**To Reproduce**
Install Ruff:
```sh
$ pip install ruff
```
Confirm that the `ruff` binary is in path and can be executed:
```sh
$ ruff --version
ruff 0.0.278
```
In `alembic.ini`:
```ini
...
[post_write_hooks]
hooks = ruff
ruff.type = console_scripts
ruff.entrypoint = ruff
ruff.options = --fix REVISION_SCRIPT_FILENAME
...
```

**Error**

```
  Generating /srv/app/alembic/versions/2023-07-15_2341_ecc9e22f7cf5_.py ...  done
  Running post write hook {name!r} ...
  FAILED
  FAILED: Could not find entrypoint console_scripts.ruff
```

**Versions.**
 - OS: Debian 12.0
 - Python: 3.11.4
 - Alembic: 1.11.1
 - SQLAlchemy: 2.0.19
 - Database: irrelevant
 - DBAPI: irrelevant

**Additional context**

**Have a nice day!**

### PR DESCRIPTION
**Pull request by lanzz**
Fixes #1275 

### Description
Implements new `exec` post-write hook runner that will execute arbitrary binaries, not just Python modules' entrypoints.

### Checklist
This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Review by CaselIT — COMMENTED** [review]
Change looks good.

not sure about the naming. spawn may be an alternative. let's wait mike opinion

File: alembic/templates/async/alembic.ini.mako [hunk 1]
```diff
@@ -72,6 +72,12 @@ sqlalchemy.url = driver://user:pass@localhost/dbname
 # black.entrypoint = black
 # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+# format using "ruff" - use the exec runner, execute a binary
+# hooks = ruff
+# ruff.type = exec
+# ruff.executable = .venv/bin/ruff
```
Comment (CaselIT) [inline alembic/templates/async/alembic.ini.mako:78]: @zzzeek can this use %(here)s to expand the current absolute path?

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 28981ecd046be0eaf54a659b2becaf4907f71a7a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 28981ecd046be0eaf54a659b2becaf4907f71a7a: https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

**Comment by zzzeek** [comment]
thanks!  I am going away near the end of this week so this may not get merged until I'm back

File: alembic/templates/async/alembic.ini.mako (hunk 1 above)
Comment (zzzeek) [reply alembic/templates/async/alembic.ini.mako:78]: sure, it should, that woudl be nice

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision a2cd43b7d4b478d805b148ad0f15055d40fb2f1f of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset a2cd43b7d4b478d805b148ad0f15055d40fb2f1f added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

**Comment by zzzeek** [comment]
any news on this?

**Comment by lanzz** [comment]
I was also away on holidays, I've updated it now to address the linter complaints. `pre-commit run --source=origin/master --origin=HEAD` and `tox -e pep8` now pass locally, not sure how that compares to the `pep8` in gerrit

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 3b1086d51fb3519e4ea9d5d7a5298c86bcc17f1e of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 3b1086d51fb3519e4ea9d5d7a5298c86bcc17f1e added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: alembic/templates/async/alembic.ini.mako [hunk 2]
```diff
@@ -72,6 +72,12 @@ sqlalchemy.url = driver://user:pass@localhost/dbname
 # black.entrypoint = black
 # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+# format using "ruff" - use the exec runner, execute a binary
```
Comment (sqla-tester) [inline alembic/templates/async/alembic.ini.mako:75]: **mike bayer** (zzzeek) wrote:

"ruff" is a linter, I see "--fix" attempts to fix things, OK we should rephrase this as "lint with attempts to fix using ruff", it doesnt seem like this is exactly a "formatter" ...thoughts?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: alembic/templates/multidb/alembic.ini.mako [hunk 3]
```diff
@@ -79,6 +79,12 @@ sqlalchemy.url = driver://user:pass@localhost/dbname2
 # black.entrypoint = black
 # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+# format using "ruff" - use the exec runner, execute a binary
```
Comment (sqla-tester) [inline alembic/templates/multidb/alembic.ini.mako:82]: **mike bayer** (zzzeek) wrote:

same doc change as we agree on w/ the other file I noted

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: docs/build/autogenerate.rst [hunk 4]
```diff
@@ -747,20 +747,29 @@ Above, we configure ``hooks`` to be a single post write hook labeled
...
 * ``type`` - this is the type of hook we are running.  Alembic includes
-  a hook runner called ``"console_scripts"``, which is specifically a
+  two hook runners: ``"console_scripts"``, which is specifically a
   Python function that uses ``subprocess.run()`` to invoke a separate
-  Python script against the revision file.  For a custom-written hook
-  function, this configuration variable would refer to the name under
+  Python script against the revision file; and ``"exec"``, which uses
+  ``subprocess.run()`` to execute an arbitrary binary.  For a custom-written
+  hook function, this configuration variable would refer to the name under
   which the custom hook was registered; see the next section for an example.
```
Comment (sqla-tester) [inline docs/build/autogenerate.rst:755]: **mike bayer** (zzzeek) wrote:

needs ".. versionadded <target version>: added new exec runner"

@caselit do we want to use major release version for this as it's totally new feature?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: docs/build/autogenerate.rst [hunk 5]
```diff
@@ -778,7 +787,7 @@
         autopep8.entrypoint = autopep8
         autopep8.options = --in-place REVISION_SCRIPT_FILENAME
 
-* ``cwd`` - optional working directory from which the console script is run.
+* ``cwd`` - optional working directory from which the formatting tool is run.
```
Comment (sqla-tester) [inline docs/build/autogenerate.rst:790]: **mike bayer** (zzzeek) wrote:

"code processing tool"

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: docs/build/tutorial.rst [hunk 6]
```diff
@@ -200,6 +200,12 @@ The file generated with the "generic" configuration looks like::
     # black.entrypoint = black
     # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+    # format using "ruff" - use the exec runner, execute a binary
```
Comment (sqla-tester) [inline docs/build/tutorial.rst:203]: **mike bayer** (zzzeek) wrote:

same doc change as we agree on w/ the other file I noted

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

**Comment by lanzz** [comment]
I've addressed all of these comments except the changelog

**Comment by sqla-tester** [comment]
Patchset 24e576b9d8aa0d3931f2b0c0c15bb370577d7485 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: alembic/templates/async/alembic.ini.mako [hunk 7]
```diff
@@ -72,6 +72,12 @@ sqlalchemy.url = driver://user:pass@localhost/dbname
 # black.entrypoint = black
 # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+# lint with attempts to fix using "ruff" - use the exec runner, execute a binary
```
Comment (sqla-tester) [inline alembic/templates/async/alembic.ini.mako:75]: **mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: alembic/templates/generic/alembic.ini.mako [hunk 8]
```diff
@@ -74,6 +74,12 @@ sqlalchemy.url = driver://user:pass@localhost/dbname
 # black.entrypoint = black
 # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+# lint with attempts to fix using "ruff" - use the exec runner, execute a binary
```
Comment (sqla-tester) [inline alembic/templates/generic/alembic.ini.mako:77]: **mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: alembic/templates/multidb/alembic.ini.mako [hunk 9]
```diff
@@ -79,6 +79,12 @@ sqlalchemy.url = driver://user:pass@localhost/dbname2
 # black.entrypoint = black
 # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+# lint with attempts to fix using "ruff" - use the exec runner, execute a binary
```
Comment (sqla-tester) [inline alembic/templates/multidb/alembic.ini.mako:82]: **mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: docs/build/autogenerate.rst [hunk 10]
```diff
@@ -730,8 +730,8 @@ Basic Formatter Configuration
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 The ``alembic.ini`` samples now include commented-out configuration
-illustrating how to configure code-formatting tools to run against the newly
-generated file path.    Example::
+illustrating how to configure code-formatting tools, or other tools like linters
```
Comment (sqla-tester) [inline docs/build/autogenerate.rst:733]: **mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: docs/build/autogenerate.rst [hunk 11]
```diff
@@ -778,7 +789,7 @@
         autopep8.entrypoint = autopep8
         autopep8.options = --in-place REVISION_SCRIPT_FILENAME
```
Comment (sqla-tester) [inline docs/build/autogenerate.rst:790]: **mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: docs/build/tutorial.rst [hunk 12]
```diff
@@ -200,6 +200,12 @@ The file generated with the "generic" configuration looks like::
     # black.entrypoint = black
     # black.options = -l 79 REVISION_SCRIPT_FILENAME
 
+    # lint with attempts to fix using "ruff" - use the exec runner, execute a binary
```
Comment (sqla-tester) [inline docs/build/tutorial.rst:203]: **mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

i will push out the changelog plus ensure 1.12

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

File: docs/build/autogenerate.rst (hunk 4 above)
Comment (sqla-tester) [reply docs/build/autogenerate.rst:755]: **mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

just a typo in the changenote

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788

* docs/build/unreleased/1275.rst (line 7): the the

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

this is a great PR, thanks for the very high quality work!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4788
